components/cdc/src/channel.rs

```diff
@@ -0,0 +1,233 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use futures::{
+    channel::mpsc::{
+        channel, unbounded, Receiver, SendError as FuturesSendError, Sender, TrySendError,
+        UnboundedReceiver, UnboundedSender,
+    },
+    stream, SinkExt, Stream, StreamExt,
+};
+use grpcio::{Result as GrpcResult, WriteFlags};
+use kvproto::cdcpb::ChangeDataEvent;
+
+use tikv_util::impl_display_as_debug;
+
+use crate::service::{CdcEvent, EventBatcher};
+
+const CDC_MSG_MAX_BATCH_SIZE: usize = 128;
+// Assume the average size of event is 1KB.
+// 2 = (CDC_MSG_MAX_BATCH_SIZE * 1KB / CDC_EVENT_MAX_BATCH_SIZE).ceil() + 1 /* reserve for ResolvedTs */;
+pub const CDC_EVENT_MAX_BATCH_SIZE: usize = 2;
+
+pub fn canal(buffer: usize) -> (Sink, Drain) {
+    let (unbounded_sender, unbounded_receiver) = unbounded();
+    let (bounded_sender, bounded_receiver) = channel(buffer);
+    (
+        Sink {
+            unbounded_sender,
+            bounded_sender,
+        },
+        Drain {
+            unbounded_receiver,
+            bounded_receiver,
+        },
+    )
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum SendError {
+    Full,
+    Disconnected,
+    Congest,
+}
+
+impl std::error::Error for SendError {}
+
+impl_display_as_debug!(SendError);
+
+macro_rules! impl_from_future_send_error {
+    ($($f:ty,)+) => {
+        $(
+            impl From<$f> for SendError {
+                fn from(e: $f) -> Self {
+                    if e.is_disconnected() {
```

Comment:
How about using pattern matching like 
```rust
match e  {
    $f::Disconnected => {...},
}
```

Reply:
`futures::channel::mpsc::{SendError, TrySendError}` are not enums. 😅

Reply:
See more: https://docs.rs/futures/0.3.14/futures/channel/mpsc/struct.SendError.html